### PR TITLE
Post training upload performance and prediction data

### PR DESCRIPTION
**Pull request by samiyaakhtar**
1. Upload model blob (re-use image container for now), getting a Model URL
2. Read performance CSV, parse and get overall average and dictionary of averages
3. Make instance of TrainingSession, upload to db and get the id by calling new train api
5. Read the prediction CSV data, convert the bounding boxes
6. Call Labels API, send the prediction data

### CODE REVIEW

File: functions/pipeline/shared/db_access/db_access_v2.py [hunk 1]
```diff
@@ -424,23 +424,23 @@ def add_training_session(self, training: TrainingSession, user_id: int):
                 cursor = conn.cursor()
                 query = ("WITH t AS ( "
                         "INSERT INTO Training_Info (TrainingDescription, ModelLocation, ClassPerfAvg, CreatedByUser) "
-                        "VALUES ({},{},{},{}) RETURNING TrainingId), "
+                        "VALUES ('{}','{}',{},{}) RETURNING TrainingId), "
```
Comment (andrebriggs) [inline functions/pipeline/shared/db_access/db_access_v2.py:427]: @samiyaakhtar thanks for adding these. I didn't get a chance to make the integration test.

File: train_vnext/training.py [hunk 2]
```diff
@@ -158,17 +142,85 @@ def process_post_training_csv(csv_path, training_id, classification_name_to_clas
...
                                     int(row[6]), 
                                     int(row[7]), 
                                     float(row[8]), 
                                     float(row[9]))
                 payload_json.append(prediction_label)
     return jsonpickle.encode(payload_json, unpicklable=False)
 
+def save_training_session(config, model_location, perf_location, prediction_labels_location):
+    # First, upload the model to blob storage
+    cur_date_time = "{date:%Y_%m_%d_%H_%M_%S}".format(date=datetime.datetime.now())
```
Comment (andrebriggs) [inline train_vnext/training.py]: @samiyaakhtar Perhaps we should prefer datetime.utcnow()

File: train_vnext/training.py [hunk 3]
```diff
@@ -158,17 +142,85 @@ def process_post_training_csv(csv_path, training_id, classification_name_to_clas
...
+    # Create a new training session in db and get its id
+    training_id = construct_new_training_session(perf_location, classification_name_to_class_id, overall_average, cur_date_time, model_location, config.get("tagging_user"), config.get("url"))
+
+    # Upload prediction labels to the db
+    upload_data_post_training(prediction_labels_location, classification_name_to_class_id, training_id, config.get("tagging_user"), config.get("url"))
+
+def upload_model_to_blob_storage(config, model_location, file_name):
+    blob_storage = BlobStorage.get_azure_storage_client(config)
+    uri = 'https://' + config.get("storage_account") + '.blob.core.windows.net/' + config.get("storage_container") + '/' + file_name
+    blob_storage.create_blob_from_path(
```
Comment (andrebriggs) [inline train_vnext/training.py:192]: @samiyaakhtar consider adding metadata to the blob like we do [here](https://github.com/CatalystCode/active-learning-detect/blob/a941ea65745320a73887f9fa6faf9769d363f541/functions/pipeline/onboardqueueproccessor/__init__.py#L77). I would include at least  training id and  user_name

File: train_vnext/training.py [hunk 4]
```diff
@@ -158,17 +142,85 @@ def process_post_training_csv(csv_path, training_id, classification_name_to_clas
...
+            config.get("storage_container"),
+            file_name,
+            model_location
+        )
+    print("Model uploaded at " + str(uri))
+    return uri
+
+def construct_new_training_session(perf_location, classification_name_to_class_id, overall_average, training_description, model_location, user_name, function_url):
+    avg_dictionary = {}
+    with open(perf_location) as csvfile:
```
Comment (andrebriggs) [inline train_vnext/training.py]: @samiyaakhtar Consider putting the reading of the perf file into it's own function.

File: train_vnext/training.py [hunk 5]
```diff
@@ -158,17 +142,85 @@ def process_post_training_csv(csv_path, training_id, classification_name_to_clas
                 prediction_label = PredictionLabel(training_id, 
                                     int(row[0].split('.')[0]), 
                                     classification_name_to_class_id[class_name], 
-                                    float(row[2]), 
-                                    float(row[3]), 
-                                    float(row[4]), 
-                                    float(row[5]), 
+                                    float(row[2])*float(row[6]),
```
Comment (andrebriggs) [inline train_vnext/training.py]: @samiyaakhtar row[6] is Height and row[7] is Width so we what the XMin and XMax to be divided by the width (x-axis) and YMin YMax to be divided by the Height (Y Axis)

**Review by andrebriggs — CHANGES_REQUESTED** [review]

File: train_vnext/training.py (hunk 5 above)
Comment (samiyaakhtar) [reply train_vnext/training.py]: good catch

File: train_vnext/training.py (hunk 4 above)
Comment (samiyaakhtar) [reply train_vnext/training.py]: will do

File: train_vnext/training.py (hunk 3 above)
Comment (samiyaakhtar) [reply train_vnext/training.py:192]: So training id we get back after we have already uploaded the blob to the storage. I can add the original file path and username maybe>?

File: train_vnext/training.py (hunk 2 above)
Comment (samiyaakhtar) [reply train_vnext/training.py]: good catch

File: train_vnext/training.py (hunk 5 above)
Comment (andrebriggs) [reply train_vnext/training.py]: @samiyaakhtar sorry I gave instructions on how to convert from absolute to relative. You're attempting to convert from relative to absolute so multiplication is correct instead of divide.

Perhaps we should have static methods on the Prediction Label that do this conversion for us. Int would be more declarative to have: 

```
prediction_label = PredictionLabel(...)
prediction_label = prediction_label.ConvertToAbsolute()
```
What do you think?

Comment (samiyaakhtar) [reply train_vnext/training.py]: yeah i think this is a better idea, let me do this as part of this PR later today

File: train_vnext/training.py [hunk 6]
```diff
@@ -155,38 +156,36 @@ def process_post_training_csv(csv_path, training_id, classification_name_to_clas
...
+    training_id = construct_new_training_session(perf_location, classification_name_to_class_id, overall_average, cur_date_time, model_location, avg_dictionary, config.get("tagging_user"), config.get("url"))
 
     # Upload prediction labels to the db
     upload_data_post_training(prediction_labels_location, classification_name_to_class_id, training_id, config.get("tagging_user"), config.get("url"))
 
-def upload_model_to_blob_storage(config, model_location, file_name):
+def upload_model_to_blob_storage(config, model_location, file_name, user_name):
     blob_storage = BlobStorage.get_azure_storage_client(config)
+    blob_metadata = {
+            "userFilePath": model_location,
```
Comment (andrebriggs) [inline train_vnext/training.py:188]: @samiyaakhtar I'm not sure how useful the userFilePath for the model on the training machine will be. I have been thinking we should save the config.ini (it will have that location as well) somewhere but that is out of scope. Storing the training id would be more useful right now

**Review by andrebriggs — CHANGES_REQUESTED** [review]

File: train_vnext/training.py (hunk 6 above)
Comment (samiyaakhtar) [reply train_vnext/training.py:188]: So training id we get back after we have already uploaded the blob to the storage. I can add the original file path and username maybe?

Comment (samiyaakhtar) [reply train_vnext/training.py:188]: @andrebriggs that's the reason I couldn't add training id to it, we have a cycle here. We want training id for the metadata but for training id we need model location. Haha

Comment (andrebriggs) [reply train_vnext/training.py:188]: @samiyaakhtar got it. My bad.

File: train_vnext/training.py (hunk 5 above)
Comment (samiyaakhtar) [reply train_vnext/training.py]: I thought it was better to place the convert to relative in the Tag class since we convert human predicted labels to relative. Let me know if you have more feedback.

File: functions/pipeline/shared/db_access/models.py [hunk 7]
```diff
@@ -84,7 +91,7 @@ def __init__(self, training_id, image_id, classification_id, x_min, x_max, y_min
         self.box_confidence = box_confidence
         self.image_confidence = image_confidence
 
-    def convert_to_relative(self):
+    def convert_to_absolute(self):
```
Comment (andrebriggs) [inline functions/pipeline/shared/db_access/models.py:94]: @samiyaakhtar So this method and the convert_to_relative on the Tag object are manipulating the current object, not returning a new object. I think returning a new object might be safer and less confusing for consumers of the object. What do you think?

Comment (samiyaakhtar) [reply functions/pipeline/shared/db_access/models.py:94]: Maybe, though it's considered good practice manipulating objects within the class itself too, if that makes readability better I will convert them to static.

Comment (samiyaakhtar) [reply functions/pipeline/shared/db_access/models.py:94]: the more I think about it, I think if it's a static function that sends a new object back it's as good as not using the class object itself. Let me know if you think I should still go ahead and make it static within the Tag and PredictionLabel class

**Review by andrebriggs — APPROVED** [review]
